### ISSUE TITLE
Generic type alias fails to resolve fields 

### ISSUE THREAD

**Issue by cart** [rust-lang/rust-analyzer#15943]
**rust-analyzer version**: 0.3.1740-standalone
**rustc version**:  1.73.0

```rust
type Wrap<T> = T;

enum X {
    A { cool: u32, stuff: u32 },
    B,
}

fn main() {
    let wrapped = Wrap::<X>::A {
        // autocomplete for fields breaks here
        cool: 100,
        stuff: 100,
    };

    // autocomplete for fields breaks here
    // RA fails to resolve `cool`'s type (lists as unknown)
    if let Wrap::<X>::A { cool, .. } = &wrapped {}
}
```

And a screenshot of the failed type hint:
![image](https://github.com/rust-lang/rust-analyzer/assets/2694663/919bced7-00c8-4875-99d5-8b81a4fb8e60)

## Context

I'm [building a DSL for Bevy Scenes](https://github.com/bevyengine/bevy/discussions/9538) with the goal of supporting as many RA niceties as possible (ex: autocomplete, go to definition, etc). I [created a discussion here a few months ago](https://github.com/rust-lang/rust-analyzer/discussions/15452) and got very helpful advice, which has generally worked!

However when building support for enums in my DSL, I encountered a limitation in the Rust type system that required the use of [this type alias wrapper workaround](https://github.com/rust-lang/rust/issues/86935) (or an experimental feature ... which isn't an option for us).

The actual proc_macro code in my impl looks like this:

```rust
 if let Wrap::<<#path as Schematic>::Props>::#variant { #(#fields,)* .. } = &mut props {
 }
```

If I replace `Wrap::<<#path as Schematic>::Props>` with the "actual" type, RA field autocomplete does work for the enum in my DSL. Sadly, I cannot name the type in the macro (as it is defined outside of the macro), so I need to rely on the type system to resolve this.

Given that RA fails to resolve the fields in the simplified `if let Wrap::<X>::A { cool, .. } = &wrapped {}`, I'm assuming that if RA can handle autocomplete for this case, it can probably handle autocompletion in the Bevy DSL proc_macro as well.

**Comment by cart** [comment]
This is the "real" case I would like to resolve. I'm guessing this is resolved by fixing the simpler case, but I'll include it here for completeness.

```rust
type Wrap<T> = T;

enum X {
    A { cool: u32, stuff: u32 },
    B,
}

enum XProps {
    A { cool: u32, stuff: u32 },
    B,
}

trait Schematic {
    type Props;
}

impl Schematic for X {
    type Props = XProps;
}

fn main() {
    let wrapped = Wrap::<<X as Schematic>::Props>::A {
        // autocomplete breaks here
        cool: 100,
        stuff: 100,
    };

    // RA resolves `cool` to unknown
    if let Wrap::<<X as Schematic>::Props>::A { cool, .. } = &wrapped {}
}
```

**Comment by Veykril** [comment]
Relevant part for this should be here I think https://github.com/rust-lang/rust-analyzer/blob/1a5cee17311fa0bb55757b382be918520b6ac2b4/crates/hir-ty/src/infer.rs#L1154-L1170

Unsure why this fails though, since `resolve_variant_on_alias` does at least handle enums https://github.com/rust-lang/rust-analyzer/blob/1a5cee17311fa0bb55757b382be918520b6ac2b4/crates/hir-ty/src/infer.rs#L1220-L1232

**Comment by cart** [comment]
I can confirm the first case now works on nightly.  The second case (which is unfortunately my Bevy experiment use case) doesn't appear to be working.

**Comment by Veykril** [comment]
Can you extract what the proc-macro generates (by looking at the expansion or similar) somehow? It's difficult to figure out a concrete failing case here otherwise

**Comment by Veykril** [comment]
Ah I think it's because we fail to resolve the generic argument which makes the entire thing not resolve anymore. So the `<#path as Schematic>::Props` thing fails to resolve most likely.

**Comment by Veykril** [comment]
```rs
type Wrap<T> = T;

struct S;

trait Schematic {
    type Props;
}

impl Schematic for S {
    type Props = X;
}

enum X {
    A { cool: u32, stuff: u32 },
    B,
}

fn main() {
    let wrapped = Wrap::<<S as Schematic>::Props>::A {
        cool: 100,
        stuff: 100,
    };

    if let Wrap::<<S as Schematic>::Props>::A { cool, ..} = &wrapped {}
                                              //^^^^ &u32
}
```
fails to resolve when it should

**Comment by Veykril** [comment]
I'm not sure whats going on but our substitutions fall back to errors once we put them in the projection here 
https://github.com/rust-lang/rust-analyzer/blob/2ee17bc5f256e379c4df954e984b9bd0d79f1cbd/crates/hir-ty/src/lower.rs#L513-L524

Debug printing the substitution on line 513 yields
```
[crates\hir-ty\src\lower.rs:520] &substitution = [?0 := AdtId(StructId(StructId(0)))<[]>]
```
whereas debug printing the projection right after yields
```
[crates\hir-ty\src\lower.rs:521] TyKind::Alias(AliasTy::Projection(ProjectionTy {
                associated_ty_id: to_assoc_type_id(associated_ty),
                substitution,
            })).intern(Interner) = AliasTy(?)
```
do you have any idea what could be going on there from a quick glance @flodiebold? Do we need to commit something somehow somewhere?

**Comment by flodiebold** [comment]
If you mean the `?` in `AliasTy(?)`, that's not an error type, it's simply the debug printing not being able to resolve the interned projection type because it doesn't have access to the database. The Chalk printing infrastructure needs the db in TLS to fully work, which is done by wrapping code in `hir_ty::tls::set_current_program`, but that's mostly not done.

**Comment by cart** [comment]
> Can you extract what the proc-macro generates (by looking at the expansion or similar) somehow? It's difficult to figure out a concrete failing case here otherwise. We most likely fail to resolve <#path as Schematic>::Props

Thats [what this chunk of code](https://github.com/rust-lang/rust-analyzer/issues/15943#issuecomment-1820273059) is. It looks like you either found it or reverse engineered the same thing though!

**Comment by Austaras** [comment]
Hmm. Let me try again.

**Comment by Austaras** [comment]
Is this piece of code workaround for rust-lang/rust#86935?
